### PR TITLE
KAFKA-13721: asymetric join-winodws should not emit spurious left/outer join results

### PR DESCRIPTION
**Pull request by mjsax**
Call for review @spena @guozhangwang

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java [hunk 1]
```diff
@@ -211,7 +214,7 @@ private void emitNonJoinedOuterRecords(final KeyValueStore<TimestampedKeyAndJoin
             // to reduce runtime cost, we try to avoid paying those cost
 
             // only try to emit left/outer join results if there _might_ be any result records
-            if (sharedTimeTracker.minTime >= sharedTimeTracker.streamTime - joinAfterMs - joinGraceMs) {
+            if (sharedTimeTracker.minTime >= sharedTimeTracker.streamTime - joinSpuriousLookBackTimeMs - joinGraceMs) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:217]: This is the actual fix.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java [hunk 2]
```diff
@@ -1197,7 +1194,7 @@ public void testAsymmetricWindowingAfter() {
         joined = stream1.join(
             stream2,
             MockValueJoiner.TOSTRING_JOINER,
-            JoinWindows.ofTimeDifferenceAndGrace(ofMillis(0), ofMillis(0)).after(ofMillis(100)),
+            JoinWindows.ofTimeDifferenceWithNoGrace(ofMillis(0)).after(ofMillis(100)),
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java:1197]: Minor side cleanup

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoinTest.java [hunk 3]
```diff
@@ -53,9 +56,14 @@
     private final String topic1 = "topic1";
     private final String topic2 = "topic2";
     private final Consumed<Integer, String> consumed = Consumed.with(Serdes.Integer(), Serdes.String());
-    private final Properties props = StreamsTestUtils.getStreamsConfig(Serdes.String(), Serdes.String());
+    private final static Properties PROPS = StreamsTestUtils.getStreamsConfig(Serdes.String(), Serdes.String());
 
-    @SuppressWarnings("deprecation")
+    @BeforeClass
+    public static void beforeClass() {
+        PROPS.put(InternalConfig.EMIT_INTERVAL_MS_KSTREAMS_OUTER_JOIN_SPURIOUS_RESULTS_FIX, 0L);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoinTest.java:63]: Side fix: we need to ensure that we never delay outputting left/outer join result due to wall-clock time not advancing using TTD

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoinTest.java [hunk 4]
```diff
@@ -749,6 +757,70 @@ public void testWindowing() {
...
+            // push one item to the other stream; this should produce one full-join item
+            // w1 = { 0:A0 (ts: 0), 1:A1 (ts: 1) }
+            // w2 = {}
+            // --> w1 = { 0:A0 (ts: 0), 1:A1 (ts: 1) }
+            // --> w2 = { 0:a0 (ts: 100) }
+            time += 100L;
+            inputTopic2.pipeInput(expectedKeys[0], "a" + expectedKeys[0], time);
+
+            processor.checkAndClearProcessResult(
+                new KeyValueTimestamp<>(0, "A0+a0", 100L)
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoinTest.java:807]: Without the fix, we get a spurious left-join result here.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoinTest.java [hunk 5]
```diff
@@ -877,7 +949,6 @@ private void testUpperWindowBound(final int[] expectedKeys,
 
         // push a dummy record to produce all left-join non-joined items
         time += 301L;
-        driver.advanceWallClockTime(Duration.ofMillis(1000L));
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamLeftJoinTest.java:880]: Don't need this any more as we disabled the wall-clock "guard" globally.

**Review by spena — APPROVED** [review]
LGTM

**Review by guozhangwang — APPROVED** [review]
Thanks @mjsax ! LGTM.

**Comment by mjsax** [comment]
Merged to `trunk` and cherry-picked to `3.1` branch.

**Pull request merged**